extra/interface/src/main/java/org/spongepowered/configurate/interfaces/InterfaceTypeSerializer.java

```diff
@@ -0,0 +1,106 @@
+import org.spongepowered.configurate.ConfigurationNode;
+import org.spongepowered.configurate.objectmapping.ConfigSerializable;
+import org.spongepowered.configurate.serialize.SerializationException;
+import org.spongepowered.configurate.serialize.TypeSerializer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.AnnotatedType;
+import java.lang.reflect.Type;
+import java.net.URL;
+import java.util.Locale;
+import java.util.Properties;
+import java.util.StringJoiner;
+
+final class InterfaceTypeSerializer implements TypeSerializer<Object> {
+
+    public static final InterfaceTypeSerializer INSTANCE = new InterfaceTypeSerializer();
+
+    private final Properties interfaceMappings = new Properties();
+
+    public static boolean applicable(final AnnotatedType type) {
+        return type.isAnnotationPresent(ConfigSerializable.class) && erase(type.getType()).isInterface();
+    }
+
+    private InterfaceTypeSerializer() {
+        final @Nullable URL mappingsUrl = getClass().getClassLoader().getResource(Constants.MAPPING_FILE);
+        if (mappingsUrl == null) {
+            return;
+        }
+
+        try (InputStream stream = mappingsUrl.openStream()) {
+            this.interfaceMappings.load(stream);
+        } catch (final IOException exception) {
+            throw new RuntimeException("Could not load interface mappings!", exception);
+        }
+    }
+
+    @Override
+    public Object deserialize(final Type type, final ConfigurationNode node) throws SerializationException {
+        final String canonicalName = erase(type).getTypeName();
+        final @Nullable String typeImpl = this.interfaceMappings.getProperty(canonicalName);
+        if (typeImpl == null) {
+            throw new SerializationException(String.format(
+                Locale.ROOT,
+                "No mapping found for type %s. Available mappings: %s",
+                canonicalName, availableMappings()
+            ));
+        }
+
+        final Class<?> implClass;
+        try {
+            implClass = Class.forName(typeImpl, true, erase(type).getClassLoader());
+        } catch (final ClassNotFoundException exception) {
+            throw new SerializationException(String.format(
+                Locale.ROOT,
+                "Could not find implementation class %s for type %s!",
+                typeImpl, canonicalName
+            ));
+        }
+
+        final @Nullable TypeSerializer<?> serializer = node.options().serializers().get(implClass);
+        if (serializer == null) {
+            throw new SerializationException("No serializer found for implementation class " + implClass);
+        }
+        return serializer.deserialize(implClass, node);
+    }
+
+    @Override
+    public void serialize(
+        final Type type,
+        final @Nullable Object obj,
+        final ConfigurationNode node
+    ) throws SerializationException {
+        throw new SerializationException(
```

Comment:
This feels like it'll cause issues, I think you have to delegate to the impl class here too.